backend/Application/Aggregates/Contract/Jobs/ParallelBatchBlockHeightJob.cs

```diff
@@ -0,0 +1,126 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Application.Aggregates.Contract.Configurations;
+using Application.Aggregates.Contract.Observability;
+using Application.Observability;
+using Microsoft.Extensions.Options;
+using Serilog.Context;
+
+namespace Application.Aggregates.Contract.Jobs;
+
+/// <summary>
+/// This class is used for jobs which depends on a block height as input. From the block height input the
+/// job will update any relevant events and update database. The import can be from both existing data in database
+/// or from the nodes.
+///
+/// This class calls the input job in batches and parallel. Hence the order of block height processed is out of order.
+/// Because of this the jobs are expected to be stateless.
+/// </summary>
+internal sealed class ParallelBatchBlockHeightJob<TStatelessJob> : IContractJob where TStatelessJob : IStatelessBlockHeightJobs
+{
+    
+    private readonly ContractHealthCheck _healthCheck;
+    private readonly IStatelessBlockHeightJobs _statelessJob;
+    private readonly ILogger _logger;
+    private readonly ContractAggregateOptions _contractAggregateOptions;
+    private readonly ContractAggregateJobOptions _jobOptions;
+
+    public ParallelBatchBlockHeightJob(
+        TStatelessJob statelessJob,
+        IOptions<ContractAggregateOptions> options,
+        ContractHealthCheck healthCheck
+        )
+    {
+        _statelessJob = statelessJob;
+        _healthCheck = healthCheck;
+        _logger = Log.ForContext<ParallelBatchBlockHeightJob<TStatelessJob>>();
+        _contractAggregateOptions = options.Value;
+        var gotJobOptions = _contractAggregateOptions.Jobs.TryGetValue(GetUniqueIdentifier(), out var jobOptions);
+        _jobOptions = gotJobOptions ? jobOptions! : new ContractAggregateJobOptions();     
+    }
+
+    /// <inheritdoc/>
+    public string GetUniqueIdentifier() => _statelessJob.GetUniqueIdentifier();
+    
+    public async Task StartImport(CancellationToken token)
+    {
+        using var _ = TraceContext.StartActivity(GetUniqueIdentifier());
```

Comment:
We can avoid the `var` part but then we will need to have an indention level added.

In C# you can have these using scopes which is active until the end of the function (like above) but then you need to have var even though the return type is void - very strange I dont’ know why. Hope they will fix one day.

I like to avoid having two much indention due to readability.